src/login/logind.c

```diff
@@ -473,27 +443,85 @@ static int manager_enumerate_sessions(Manager *m) {
 
-                RET_GATHER(r, session_load(s));
+                session_add_to_gc_queue(s);
         }
 
-        /* We might be restarted and PID1 could have sent us back the session device fds we previously
-         * saved. */
-        (void) manager_attach_fds(m);
+        n = sd_listen_fds_with_names(true, &fdnames);
+        if (n < 0)
+                return log_error_errno(n, "Failed to acquire passed fd list: %m");
 
-        return r;
+        for (int i = 0; i < n; i++) {
+                int fd = SD_LISTEN_FDS_START + i;
+                _cleanup_free_ char *id = NULL;
+                dev_t devno;
+                Session *s;
+
+                r = parse_fdname(fdnames[i], &id, &devno);
+                if (r < 0) {
+                        RET_GATHER(ret, log_warning_errno(r,
+                                                          "Failed to parse stored fd name '%s', ignoring: %m",
+                                                          fdnames[i]));
+                        goto fail_skip;
+                }
+
+                s = hashmap_get(m->sessions, id);
+                if (!s) {
+                        /* If the session doesn't exist anymore, let's simply close this fd. */
+                        RET_GATHER(ret, log_debug_errno(SYNTHETIC_ERRNO(ENXIO),
+                                                        "Cannot attach fd '%s' to unknown session, ignoring: %s",
+                                                        fdnames[i], id));
+                        goto fail_skip;
+                }
+
+                if (devno > 0) {
+                        r = deliver_session_device_fd(s, fdnames[i], fd, devno);
+                        if (r < 0) {
+                                RET_GATHER(ret, r);
+                                goto fail_skip;
+                        }
+                } else {
+                        _cleanup_(pidref_done) PidRef leader_fdstore = PIDREF_NULL;
+
+                        assert(!pidref_is_set(&s->leader));
+
+                        r = pidref_set_pidfd_take(&leader_fdstore, fd);
+                        if (r < 0) {
+                                RET_GATHER(ret, log_warning_errno(r,
+                                                                  "Failed to create reference to leader of session '%s', ignoring: %m",
+                                                                  id));
+                                goto fail_skip;
+                        }
+
+                        r = session_set_leader_consume(s, TAKE_PIDREF(leader_fdstore));
+                        if (r < 0)
+                                RET_GATHER(ret, log_warning_errno(r,
+                                                                  "Failed to attach leader pidfd for session '%s', ignoring: %m",
+                                                                  id));
+
+                        // FIXME: Maybe call session_watch_pidfd() here?
```

Comment:
So I'm curious why currently this is called in `session_send_create_reply` instead of immediately in `session_set_leader_consume` 🤔

cc @msekletar